### PR TITLE
74 convert offchainassetreceiptvaulttestts tests checking confiscate to foundry test

### PR DESCRIPTION
**Pull request by ninokeldishvili**
## Motivation
convert tests checking confiscate to foundry test

## Solution
write new tests in foundry

## Checks

By submitting this for review, I'm confirming I've done the following:
- [x] made this PR as small as possible
- [x] unit-tested any new functionality
- [x] linked any relevant issues or PRs

### CODE REVIEW

File: test/foundry/Confiscate.t.sol [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+        vault.grantRole(vault.CONFISCATOR(), alice);
+
+        // Stop recording logs
+        Vm.Log[] memory logs = vm.getRecordedLogs();
+        vault.confiscateShares(bob, justification);
+
+        // Check the logs to ensure event is not present
+        bool eventFound = false;
+
+        for (uint256 i = 0; i < logs.length; i++) {
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:47]: expect emit

File: test/foundry/Confiscate.t.sol [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        shareRatio = bound(shareRatio, 1, 1e18);
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyAlice = bound(fuzzedKeyAlice, 1, SECP256K1_ORDER - 1);
+        address alice = vm.addr(fuzzedKeyAlice);
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyBob = bound(fuzzedKeyBob, 1, SECP256K1_ORDER - 1);
+        address bob = vm.addr(fuzzedKeyBob);
+
+        referenceBlockNumber = bound(referenceBlockNumber, 1, block.number);
+        certifyUntil = bound(certifyUntil, 1, 1e6 - 1);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:79]: why 1e6 - 1?

File: test/foundry/Confiscate.t.sol [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+    ) external {
+        shareRatio = bound(shareRatio, 1, 1e18);
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyAlice = bound(fuzzedKeyAlice, 1, SECP256K1_ORDER - 1);
+        address alice = vm.addr(fuzzedKeyAlice);
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyBob = bound(fuzzedKeyBob, 1, SECP256K1_ORDER - 1);
+        address bob = vm.addr(fuzzedKeyBob);
+
+        referenceBlockNumber = bound(referenceBlockNumber, 1, block.number);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:78]: do we need a warp somewhere?

File: test/foundry/Confiscate.t.sol [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        vm.assume(alice != bob);
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+        OffchainAssetReceiptVault vault = OffchainAssetVaultCreator.createVault(factory, alice, assetName, assetSymbol);
+        vault.grantRole(vault.CONFISCATOR(), alice);
+        vault.grantRole(vault.DEPOSITOR(), alice);
+        vault.grantRole(vault.CERTIFIER(), alice);
+
+        // Call the certify function
+        vault.certify(certifyUntil, referenceBlockNumber, false, justification);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:90]: why hardcoded false?

File: test/foundry/Confiscate.t.sol [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+        //set upperBound for assets so it does not overflow while calculating fixedPointDiv or fixedPointMul
+        uint256 upperBound = type(uint256).max / 1e18;
+        // Assume that aliceAssets is less than totalSupply
+        aliceAssets = bound(aliceAssets, 1, upperBound);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:95]: why not 0?

File: test/foundry/Confiscate.t.sol [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+
+        assertFalse(eventFound, "ConfiscateShares event should not be emitted");
+        vm.stopPrank();
+    }
+
+    /// Test to checks ConfiscateShares
+    function testConfiscateShares(
+        uint256 fuzzedKeyAlice,
+        uint256 fuzzedKeyBob,
+        uint256 shareRatio,
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:62]: min share ratio?

File: test/foundry/Confiscate.t.sol [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        vm.assume(alice != bob);
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+        OffchainAssetReceiptVault vault = OffchainAssetVaultCreator.createVault(factory, alice, assetName, assetSymbol);
+        vault.grantRole(vault.CONFISCATOR(), alice);
+        vault.grantRole(vault.DEPOSITOR(), alice);
+        vault.grantRole(vault.CERTIFIER(), alice);
+
+        // Call the certify function
+        vault.certify(certifyUntil, referenceBlockNumber, false, justification);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:138]: why false?

File: test/foundry/Confiscate.t.sol [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        shareRatio = bound(shareRatio, 1, 1e18);
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyAlice = bound(fuzzedKeyAlice, 1, SECP256K1_ORDER - 1);
+        address alice = vm.addr(fuzzedKeyAlice);
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyBob = bound(fuzzedKeyBob, 1, SECP256K1_ORDER - 1);
+        address bob = vm.addr(fuzzedKeyBob);
+
+        referenceBlockNumber = bound(referenceBlockNumber, 1, block.number);
+        certifyUntil = bound(certifyUntil, 1, 1e6 - 1);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:127]: why 1e6 - 1?

File: test/foundry/Confiscate.t.sol [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+    ) external {
+        shareRatio = bound(shareRatio, 1, 1e18);
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyAlice = bound(fuzzedKeyAlice, 1, SECP256K1_ORDER - 1);
+        address alice = vm.addr(fuzzedKeyAlice);
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyBob = bound(fuzzedKeyBob, 1, SECP256K1_ORDER - 1);
+        address bob = vm.addr(fuzzedKeyBob);
+
+        referenceBlockNumber = bound(referenceBlockNumber, 1, block.number);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:126]: do we need to warp?

File: test/foundry/Confiscate.t.sol [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+        vault.grantRole(vault.DEPOSITOR(), alice);
+        vault.grantRole(vault.CERTIFIER(), alice);
+
+        // Call the certify function
+        vault.certify(certifyUntil, referenceBlockNumber, false, justification);
+
+        //set upperBound for assets so it does not overflow while calculating fixedPointDiv or fixedPointMul
+        uint256 upperBound = type(uint256).max / 1e18;
+        // Assume that aliceAssets is less than totalSupply
+        aliceAssets = bound(aliceAssets, 1, upperBound);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:143]: why not 0?

File: test/foundry/Confiscate.t.sol [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+        vm.assume(alice != bob);
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+        OffchainAssetReceiptVault vault = OffchainAssetVaultCreator.createVault(factory, alice, assetName, assetSymbol);
+
+        vault.grantRole(vault.CONFISCATOR(), alice);
+
+        // Stop recording logs
+        Vm.Log[] memory logs = vm.getRecordedLogs();
+        vault.confiscateReceipt(bob, 1, justification);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:178]: why 1?

File: test/foundry/Confiscate.t.sol [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+
+        // Check the logs to ensure event is not present
+        bool eventFound = false;
+
+        for (uint256 i = 0; i < logs.length; i++) {
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:183]: expect emit

File: test/foundry/Confiscate.t.sol [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+        uint256 fuzzedKeyBob,
+        uint256 shareRatio,
+        uint256 aliceAssets,
+        string memory assetName,
+        string memory assetSymbol,
+        bytes memory justification,
+        uint256 certifyUntil,
+        uint256 referenceBlockNumber
+    ) external {
+        shareRatio = bound(shareRatio, 1, 1e18);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:206]: why not 0?

File: test/foundry/Confiscate.t.sol [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+
+        assertFalse(eventFound, "ConfiscateReceipt event should not be emitted");
+        vm.stopPrank();
+    }
+
+    /// Test to checks ConfiscateReceipt
+    function testConfiscateReceipt(
+        uint256 fuzzedKeyAlice,
+        uint256 fuzzedKeyBob,
+        uint256 shareRatio,
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:198]: min

File: test/foundry/Confiscate.t.sol [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+        shareRatio = bound(shareRatio, 1, 1e18);
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyAlice = bound(fuzzedKeyAlice, 1, SECP256K1_ORDER - 1);
+        address alice = vm.addr(fuzzedKeyAlice);
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyBob = bound(fuzzedKeyBob, 1, SECP256K1_ORDER - 1);
+        address bob = vm.addr(fuzzedKeyBob);
+
+        referenceBlockNumber = bound(referenceBlockNumber, 1, block.number);
+        certifyUntil = bound(certifyUntil, 1, 1e6 - 1);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:215]: why 1e6 - 1?

File: test/foundry/Confiscate.t.sol [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+    ) external {
+        shareRatio = bound(shareRatio, 1, 1e18);
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyAlice = bound(fuzzedKeyAlice, 1, SECP256K1_ORDER - 1);
+        address alice = vm.addr(fuzzedKeyAlice);
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyBob = bound(fuzzedKeyBob, 1, SECP256K1_ORDER - 1);
+        address bob = vm.addr(fuzzedKeyBob);
+
+        referenceBlockNumber = bound(referenceBlockNumber, 1, block.number);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:214]: do we need a warp?

File: test/foundry/Confiscate.t.sol [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+        vm.assume(alice != bob);
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+        OffchainAssetReceiptVault vault = OffchainAssetVaultCreator.createVault(factory, alice, assetName, assetSymbol);
+        vault.grantRole(vault.CONFISCATOR(), alice);
+        vault.grantRole(vault.DEPOSITOR(), alice);
+        vault.grantRole(vault.CERTIFIER(), alice);
+
+        // Call the certify function
+        vault.certify(certifyUntil, referenceBlockNumber, false, justification);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:226]: why false?

File: test/foundry/Confiscate.t.sol [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+
+        //set upperBound for assets so it does not overflow while calculating fixedPointDiv or fixedPointMul
+        uint256 upperBound = type(uint256).max / 1e18;
+        // Assume that aliceAssets is less than totalSupply
+        aliceAssets = bound(aliceAssets, 1, upperBound);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:231]: why 1?

File: test/foundry/Confiscate.t.sol [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+
+        vault.deposit(aliceAssets, bob, shareRatio, justification);
+
+        vm.expectEmit(true, true, true, true);
+        emit ConfiscateReceipt(alice, bob, 1, aliceAssets, justification);
+
+        vault.confiscateReceipt(bob, 1, justification);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:238]: why 1?

File: test/foundry/Confiscate.t.sol [hunk 20]
```diff
@@ -0,0 +1,290 @@
...
+        uint256 fuzzedKeyBob,
+        uint256 shareRatio,
+        uint256 aliceAssets,
+        string memory assetName,
+        string memory assetSymbol,
+        bytes memory justification,
+        uint256 certifyUntil,
+        uint256 referenceBlockNumber
+    ) external {
+        shareRatio = bound(shareRatio, 1, 1e18);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:254]: why not 0?

File: test/foundry/Confiscate.t.sol [hunk 21]
```diff
@@ -0,0 +1,290 @@
...
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyAlice = bound(fuzzedKeyAlice, 1, SECP256K1_ORDER - 1);
+        address alice = vm.addr(fuzzedKeyAlice);
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        fuzzedKeyBob = bound(fuzzedKeyBob, 1, SECP256K1_ORDER - 1);
+        address bob = vm.addr(fuzzedKeyBob);
+
+        referenceBlockNumber = bound(referenceBlockNumber, 1, block.number);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:262]: warp?

File: test/foundry/Confiscate.t.sol [hunk 22]
```diff
@@ -0,0 +1,290 @@
...
+        certifyUntil = bound(certifyUntil, 1, 1e6 - 1);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:263]: why 1e6 - 1?

File: test/foundry/Confiscate.t.sol [hunk 23]
```diff
@@ -0,0 +1,290 @@
...
+
+        vault.confiscateReceipt(bob, 1, justification);
+        vm.stopPrank();
+    }
+
+    /// Test to checks ConfiscatedReceipt amount is transferred
+    function testConfiscatedReceiptIsTransferred(
+        uint256 fuzzedKeyAlice,
+        uint256 fuzzedKeyBob,
+        uint256 shareRatio,
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:246]: min?

File: test/foundry/Confiscate.t.sol [hunk 24]
```diff
@@ -0,0 +1,290 @@
...
+        vm.assume(alice != bob);
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+        OffchainAssetReceiptVault vault = OffchainAssetVaultCreator.createVault(factory, alice, assetName, assetSymbol);
+        vault.grantRole(vault.CONFISCATOR(), alice);
+        vault.grantRole(vault.DEPOSITOR(), alice);
+        vault.grantRole(vault.CERTIFIER(), alice);
+
+        // Call the certify function
+        vault.certify(certifyUntil, referenceBlockNumber, false, justification);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:274]: why false?

File: test/foundry/Confiscate.t.sol [hunk 25]
```diff
@@ -0,0 +1,290 @@
...
+
+        //set upperBound for assets so it does not overflow while calculating fixedPointDiv or fixedPointMul
+        uint256 upperBound = type(uint256).max / 1e18;
+        // Assume that aliceAssets is less than totalSupply
+        aliceAssets = bound(aliceAssets, 1, upperBound);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:279]: why 1?

File: test/foundry/Confiscate.t.sol [hunk 26]
```diff
@@ -0,0 +1,290 @@
...
+
+        vault.deposit(aliceAssets, bob, shareRatio, justification);
+        vault.deposit(aliceAssets, bob, shareRatio, justification);
+
+        vm.expectEmit(true, true, true, true);
+        emit TransferSingle(address(vault), bob, alice, 1, aliceAssets);
+
+        vault.confiscateReceipt(bob, 1, justification);
```
Comment (thedavidmeister) [inline test/foundry/Confiscate.t.sol:287]: why 1?

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

File: test/foundry/Confiscate.t.sol (hunk 5 above)
Comment (ninokeldishvili) [reply test/foundry/Confiscate.t.sol]: I'm depositing alliceAssets. if its 0 will get ZeroAssetsAmount() error. this case is tested in deposits tests

File: test/foundry/Confiscate.t.sol (hunk 10 above)
Comment (ninokeldishvili) [reply test/foundry/Confiscate.t.sol]: I'm depositing alliceAssets. if its 0 will get ZeroAssetsAmount() error. this case is tested in deposits tests

File: test/foundry/Confiscate.t.sol (hunk 1 above)
Comment (ninokeldishvili) [reply test/foundry/Confiscate.t.sol:47]: actually i am checking here that the event is not emitted,

File: test/foundry/Confiscate.t.sol (hunk 11 above)
Comment (ninokeldishvili) [reply test/foundry/Confiscate.t.sol]: I do one deposit and have one receipt with id =1, If i fuzz id here, and the id is different than 1, there will be zero balance and event will not be emitted.

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

File: test/foundry/src/concrete/vault/OffchainAssetReceiptVault.confiscate.t.sol [hunk 27]
```diff
@@ -0,0 +1,271 @@
...
+pragma solidity =0.8.25;
+
+import {
+    OffchainAssetReceiptVault,
+    OffchainAssetReceiptVaultConfig
+} from "../../../../../contracts/concrete/vault/OffchainAssetReceiptVault.sol";
+import {OffchainAssetReceiptVaultTest, Vm} from "test/foundry/abstract/OffchainAssetReceiptVaultTest.sol";
+import {Receipt as ReceiptContract} from "../../../../../contracts/concrete/receipt/Receipt.sol";
+
+contract Confiscate is OffchainAssetReceiptVaultTest {
```
Comment (thedavidmeister) [inline test/foundry/src/concrete/vault/OffchainAssetReceiptVault.confiscate.t.sol:11]: `ConfiscateTest` is a better name for a test contract

File: test/foundry/src/concrete/vault/OffchainAssetReceiptVault.confiscate.t.sol [hunk 28]
```diff
@@ -0,0 +1,271 @@
...
+        assertTrue(eventFound, "OffchainAssetReceiptVaultInitialized event log not found");
+
+        // Create an receipt contract
+        ReceiptContract receipt = ReceiptContract(receiptAddress);
+
+        return receipt;
+    }
+
+    /// Checks that confiscateShares balances don't change or do change as expected
+    function checkConfiscateShares(OffchainAssetReceiptVault vault, address alice, address bob, bytes memory data)
```
Comment (thedavidmeister) [inline test/foundry/src/concrete/vault/OffchainAssetReceiptVault.confiscate.t.sol:44]: let's have one test file for share and one for receipts

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

File: test/foundry/src/concrete/vault/OffchainAssetReceiptVault.confiscateShares.t.sol [hunk 29]
```diff
@@ -0,0 +1,157 @@
...
+
+        vm.assume(alice != bob);
+
+        OffchainAssetReceiptVault vault = createVault(alice, assetName, assetSymbol);
+
+        // Prank as Alice to set role
+        vm.startPrank(alice);
+        vault.grantRole(vault.CONFISCATOR(), bob);
+        vault.grantRole(vault.DEPOSITOR(), bob);
+        // vm.stopPrank();
```
Comment (thedavidmeister) [inline test/foundry/src/concrete/vault/OffchainAssetReceiptVault.confiscateShares.t.sol:95]: commented out code here

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

**Review by thedavidmeister — APPROVED** [review]

**Pull request merged**